File: Striver/Trees/Q22.py
# print root to node path in binary tree

from Node import Node

def printPath(root, arr, target):
    if root == None:
        return False
    arr.append(root.data)
    if root.data == target:
        return True
    if printPath(root.left, arr, target) or printPath(root.right, arr, target):
        return True
    arr.pop()
    return False

root = Node(10)
root.left = Node(20)
root.right = Node(30)
root.left.left = Node(40)
arr = []
printPath(root, arr, 40)
print(arr)

File: Striver/Trees/Q22.java
package Striver.Trees;

import java.util.*;

public class Q22{
    // Print root to node path in binary tree

    public static boolean printPath(Node root, int target, ArrayList<Integer> path){
        if(root == null) return false;
        path.add(root.data);
        if(root.data == target) return true;
        if(printPath(root.left, target, path) || printPath(root.right, target, path)) return true;
        path.remove(path.size() - 1);
        return false;
    }

    public static void main(String[] args) {
        Q22 q = new Q22();
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        ArrayList<Integer> path = new ArrayList<>();
        System.out.println(printPath(root, 3, path));
        System.out.println(path);
    }
}